new-docs/content/any/getting-started/policies/index.md

```diff
@@ -9,325 +9,438 @@ description: Learn about writing Oso policies - the source of truth for authoriz
+...
+Application error: 'str' object has no attribute 'admin'
+```
 
-{{% exampleGet "userClass" %}}
+To guard against this, we can add explicit type checks using
+the Polar `matches` operator:
 
-The code above provides a `User` object as the _actor_ for our `allow` rule.
-Since `User` has a field called `{{% exampleGet "isAdmin" %}}`, it is checked
-during evaluation of the Polar rule and found to be true.
+```polar
+allow(user, "read", resource) if
+    actor matches User and
+    user.admin = true;
+
+allow(user, "read", resource) if
+    user matches User and
+    resource matches Resource and
+    user.id = resource.owner;
+```
 
-In addition to accessing attributes, you can also call methods on application
-instances in a policy:
+The `matches` operator succeeds when its left-hand side is an
+instance of (a subclass of) its right-hand side. For our example
+query above, since `str` is not a subclass of either `User`
+or `Resource`, the `matches` would fail, and not raise an error.
 
-{{< code file="policy.polar" >}}
-allow(actor, action, resource) if actor.{{% exampleGet "isAdminOf" %}}(resource);
-{{< /code >}}
+Here `User` and `Resource` refer to the application classes defined
+above. But in order for Polar to use a class in a type check, it must
+first be _registered_. If you're using an ORM adapter, this happens
+automatically for all model classes; otherwise, you can register
+classes manually using `Oso.register_class`. If you forget, Polar
+will warn you about an "unknown specializer" at load time.
 
-### Registering Application Types
+### Specializers
 
-Instances of application types can be constructed from inside an Oso policy
-using [the `new` operator](polar-syntax#new) if the class has been
-**registered**. {{% exampleGet "registerClass" %}}
+Since type checks are common and helpful for rule parameters,
+Polar provides a shortcut syntax:
 
-Once the class is registered, we can make a `User` object in Polar. This can be
-helpful for writing inline test queries:
+```polar
+allow(user: User, "read", resource: Resource) if
+    user.admin = true;
 
-{{% exampleGet "testQueries" %}}
+allow(user: User, "read", resource: Resource) if
+    user.id = resource.owner;
+```
 
-Registering classes also makes it possible to use
-[specialization](polar-syntax#specialization) and [the `matches`
-operator](polar-syntax#matches-operator) with the registered class.
+These rules mean the same thing as the ones above, but their
+first and third parameters are _specialized_ on the `User`
+and `Resource` classes, respectively: they will only match
+when the supplied arguments are instances of (subclasses of)
+those classes.
 
-In our previous example, the **allow** rule expected the actor to be a `User`,
-but we couldn’t actually check that type assumption in the policy. If we
-register the `User` class, we can write the following rule:
+It's considered good practice to use a specializer on any
+parameter whose fields you access.
```

Comment:
In the future, it would be great to have a "next level" guide that goes into cool stuff like field-level specializers. We don't have any great docs on this specific thing now, but maybe we can add a TODO here to link to them once we have it. 

Or, we could add one sentence explaining it here if you're up for it. E.g., "pattern matching can go beyond simple type checks as well. Field-level specializers can be especially useful for ensuring certain fields exist before entering a rule body:
```
allow(user: User{id: user_id}, "read", resource: Resource{owner: resource_owner}) if
    user_id = resource_owner;
```
For more information on pattern matching in Polar, see the [syntax guide](polar-syntax#rules)."